[CI:DOCS] Cirrus: Ability to skip most tests for docs updates

cevich opened the pull request:
***Warning***: `skip` has non-obvious side-effects vs `only_if`:
https://cirrus-ci.org/guide/writing-tasks/#conditional-task-execution

The skip instruction can give a false sense of security by always
marking tasks as passed in the UI, even if they didn't run. In
contrast, the `only_if` condition will avoid creating the task
all -together; therefore, a problematic task's absense is more likely to
be noticed if it introduced a problem.

***Note:*** I left the OSX build in because it actually does some platform-specific docs-manipulation we want exercised.

mheon commented:
LGTM

TomSweeneyRedHat commented:
Is this taking over for #7797?  Looks like it's working better.

edsantiago commented:
Looks like we collided. I went with `skip` because it was easy for my map maker to parse and because positive logic (`skip if PR =~ CI:DOCS`) is more readable and maintainable than negative (`only if PR !~ CI:DOCS`).

edsantiago commented on .cirrus.yml:
```diff
@@ -143,6 +143,7 @@ smoke_task:
 build_task:
     alias: 'build'
     name: 'Build for $DISTRO_NV'
+    only_if: $CIRRUS_CHANGE_TITLE !=~ '.*CI:DOCS.*'
```
Would you consider making this a macro, e.g. `&not_docs`, for readability (and DRY) below?

cevich replied:
good idea

edsantiago commented:
LGTM, modulo one request for DRY. I like how this approach doesn't have all the nasty gray bullet clutter that `skip` does. Map attached for those who find it useful for review.
![cirrus-map-pr7998](https://user-images.githubusercontent.com/265639/95790870-71963400-0c9d-11eb-977b-1d08783fe26f.png)

cevich commented:
> I went with skip because it was easy for my map maker to parse and because positive logic

@edsantiago warning...danger..danger: `skip` has non-obvious side-effects vs `only_if`:
https://cirrus-ci.org/guide/writing-tasks/#conditional-task-execution

The `skip` instruction can give a false sense of security by always marking tasks as passed in the UI, even if they didn't run.  Therefore, this includes if the PR actually breaks something in one of those green-but-skipped tasks.

Consider the position of some future CI-failure analyst:  They must go trolling back through multiple CI builds, trying to find the one that broke something.  If the culprit was a `[CI:DOCS]` task, they are much more likely to notice if it did **not** run the broken task, as opposed to seeing **all tasks are all green**.   That's why I deliberately chose to use `only_if` over `skip`, the former doesn't create the task at all, the later does and then lies about it's status.

cevich commented:
Rebased and updated w/ Ed's anchor/alias suggestion.  I also updated the commit and PR message to make clear the reasoning behind `only_if` vs `skip` usage.

cevich commented:
Clarification: What I mean by false-success status can be seen in this PRs twin, in the WebUI: https://cirrus-ci.com/build/6731773184835584

edsantiago commented:
/lgtm
/hold

cevich commented:
@edsantiago is there some reason why we're holding on this? Did I forget something important again?

edsantiago commented:
No reason to hold, I just use that convention to allow you to hold-cancel or to solicit further review; i.e. to leave that decision up to you.

cevich commented:
Oh whew! okay, thanks.

cevich commented:
/hold cancel
